<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Jason Serviss" />

<meta name="date" content="2018-04-13" />

<title>Typical workflow for filtering counts data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceLine, a.sourceLine { display: inline-block; min-height: 1.25em; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; }
@media print {
code.sourceCode { white-space: pre-wrap; }
div.sourceLine, a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource div.sourceLine, .numberSource a.sourceLine
  { position: relative; }
pre.numberSource div.sourceLine::before, .numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em; }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; color: #aaaaaa;  padding-left: 4px; }
@media screen {
a.sourceLine::before { text-decoration: underline; color: initial; }
}
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.bn { color: #40a070; } /* BaseN */
code span.fl { color: #40a070; } /* Float */
code span.ch { color: #4070a0; } /* Char */
code span.st { color: #4070a0; } /* String */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.ot { color: #007020; } /* Other */
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.fu { color: #06287e; } /* Function */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code span.cn { color: #880000; } /* Constant */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.ss { color: #bb6688; } /* SpecialString */
code span.im { } /* Import */
code span.va { color: #19177c; } /* Variable */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.op { color: #666666; } /* Operator */
code span.bu { } /* BuiltIn */
code span.ex { } /* Extension */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.at { color: #7d9029; } /* Attribute */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Typical workflow for filtering counts data</h1>
<h4 class="author"><em>Jason Serviss</em></h4>
<h4 class="date"><em>2018-04-13</em></h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The vignette is designed to show a typical pipeline for filtering counts data. Example data mirroring output from HTSeq is included and called “testingCounts” and this will be used for demonstration purposes.</p>
<pre class="sourceCode r" id="cb1"><code class="sourceCode r"><div class="sourceLine" id="cb1-1" data-line-number="1">testingCounts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##                       HGN A.htseq B.htseq C.htseq D.htseq
## 1                    ACTB      63      19      59      14
## 2                       a      36      47      28      21
## 3                       b       5       6       1      13
## 4                       c       1      11       1       9
## 5                       d      84       0       0       8
## 6                       e       8       1      20       3
## 7                       f       9      35      50       7
## 8                       g       5       0       7       1
## 9                       h       7       1       5       8
## 10                      i      17       1      17      23
## 11                      j       5       3       3      12
## 12                      k       0       1       1       0
## 13                 ERCC-1       1       0       0       2
## 14 __alignment_not_unique       1       1       0       2</code></pre>
<div id="move-gene-name-column-to-rownames" class="section level3">
<h3>Move gene name column to rownames</h3>
<p>The column called HGN, indicating the gene names is included in HTSeq counts file but we typically want gene names as rownames. The function <code>moveGenesToRownames</code> accomplishes this and returns the modified data.frame.</p>
<pre class="sourceCode r" id="cb3"><code class="sourceCode r"><div class="sourceLine" id="cb3-1" data-line-number="1">counts &lt;-<span class="st"> </span><span class="kw">moveGenesToRownames</span>(testingCounts)</div>
<div class="sourceLine" id="cb3-2" data-line-number="2">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##                        A.htseq B.htseq C.htseq D.htseq E.htseq
## ACTB                        63      19      59      14       0
## a                           36      47      28      21       1
## b                            5       6       1      13       5
## c                            1      11       1       9      12
## d                           84       0       0       8      14
## e                            8       1      20       3       0
## f                            9      35      50       7       9
## g                            5       0       7       1       8
## h                            7       1       5       8       8
## i                           17       1      17      23      47
## j                            5       3       3      12      18
## k                            0       1       1       0       7
## ERCC-1                       1       0       0       2       9
## __alignment_not_unique       1       1       0       2       3</code></pre>
</div>
<div id="remove-.htseq-suffix-from-sample-names" class="section level3">
<h3>Remove .htseq suffix from sample names</h3>
<p>HTSeq adds the suffix “.htseq” to column names when it reports counts. This function removes that suffix from the column names of the supplied counts data.frame.</p>
<pre class="sourceCode r" id="cb5"><code class="sourceCode r"><div class="sourceLine" id="cb5-1" data-line-number="1">counts &lt;-<span class="st"> </span><span class="kw">removeHTSEQsuffix</span>(counts)</div>
<div class="sourceLine" id="cb5-2" data-line-number="2">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##                         A  B  C  D  E
## ACTB                   63 19 59 14  0
## a                      36 47 28 21  1
## b                       5  6  1 13  5
## c                       1 11  1  9 12
## d                      84  0  0  8 14
## e                       8  1 20  3  0
## f                       9 35 50  7  9
## g                       5  0  7  1  8
## h                       7  1  5  8  8
## i                      17  1 17 23 47
## j                       5  3  3 12 18
## k                       0  1  1  0  7
## ERCC-1                  1  0  0  2  9
## __alignment_not_unique  1  1  0  2  3</code></pre>
</div>
<div id="extract-ercc-reads" class="section level3">
<h3>Extract ERCC reads</h3>
<p>ERCC reads, when present in the experiment, are included in HTSeq output as genes although, we typically want these to be in a seperate data structure. The <code>detectERCCreads</code> function searches the rownames for gene names (the rownames of the input data.frame) matching the specified regular expression and returns a logical indicating where they are found. The default regular expression for matching is “^ERCC\-[0-9]*$“. The returned logical vector can be subsequently used to subset the ERCC reads from the counts matrix and vice versa. Typically, we expect 92 ERCC reads to be identified and, if they are not, a warning is issued.</p>
<pre class="sourceCode r" id="cb7"><code class="sourceCode r"><div class="sourceLine" id="cb7-1" data-line-number="1">ercc &lt;-<span class="st"> </span><span class="kw">detectERCCreads</span>(counts)</div></code></pre>
<pre><code>## Warning in detectERCCreads(counts): Couldn't detect all ERCC reads.</code></pre>
<pre class="sourceCode r" id="cb9"><code class="sourceCode r"><div class="sourceLine" id="cb9-1" data-line-number="1">countsERCC &lt;-<span class="st"> </span>counts[ercc, ]</div>
<div class="sourceLine" id="cb9-2" data-line-number="2">counts &lt;-<span class="st"> </span>counts[<span class="op">!</span>ercc, ]</div>
<div class="sourceLine" id="cb9-3" data-line-number="3">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##                         A  B  C  D  E
## ACTB                   63 19 59 14  0
## a                      36 47 28 21  1
## b                       5  6  1 13  5
## c                       1 11  1  9 12
## d                      84  0  0  8 14
## e                       8  1 20  3  0
## f                       9 35 50  7  9
## g                       5  0  7  1  8
## h                       7  1  5  8  8
## i                      17  1 17 23 47
## j                       5  3  3 12 18
## k                       0  1  1  0  7
## __alignment_not_unique  1  1  0  2  3</code></pre>
<pre class="sourceCode r" id="cb11"><code class="sourceCode r"><div class="sourceLine" id="cb11-1" data-line-number="1">countsERCC[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##        A B C D E
## ERCC-1 1 0 0 2 9</code></pre>
</div>
<div id="remove-htseq-statistics" class="section level3">
<h3>Remove HTSeq statistics</h3>
<p>HTSeq typically reports the following information in the last rows of the counts data: “__no_feature“,”__ambiguous“,”__too_low_aQual“,”__not_aligned“,”__alignment_not_unique“. Although this is useful information it is not usually desired to retain this in downstream processing. The <code>detectNonGenes</code> function finds these rows and returns a logical vector allowing them to be removed.</p>
<pre class="sourceCode r" id="cb13"><code class="sourceCode r"><div class="sourceLine" id="cb13-1" data-line-number="1">counts &lt;-<span class="st"> </span>counts[<span class="op">!</span><span class="kw">detectNonGenes</span>(counts), ]</div>
<div class="sourceLine" id="cb13-2" data-line-number="2">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##       A  B  C  D  E
## ACTB 63 19 59 14  0
## a    36 47 28 21  1
## b     5  6  1 13  5
## c     1 11  1  9 12
## d    84  0  0  8 14
## e     8  1 20  3  0
## f     9 35 50  7  9
## g     5  0  7  1  8
## h     7  1  5  8  8
## i    17  1 17 23 47
## j     5  3  3 12 18
## k     0  1  1  0  7</code></pre>
</div>
<div id="remove-low-quality-genes" class="section level3">
<h3>Remove low quality genes</h3>
<p>In gene expression counts data if can often be the case that some genes are not detected. This can simply be due to the fact that the gene is not expressed in the tissue or, in addition, that the sequencing depth was not sufficient to detect the gene. In addition, some genes may be detected but in so few samples or at such a low level that it makes the quantified value highly unreliable. In these cases, it is desireable to remove the gene before downstream analysis which is facilitated by the <code>detectLowQualityGenes</code> function. A message is issued indicating the number of low quality genes detected and the percentage of low quality genes for the dataset.</p>
<pre class="sourceCode r" id="cb15"><code class="sourceCode r"><div class="sourceLine" id="cb15-1" data-line-number="1">counts &lt;-<span class="st"> </span>counts[<span class="kw">detectLowQualityGenes</span>(counts, <span class="dv">18</span>), ]</div></code></pre>
<pre><code>## [1] &quot;Detected 1 low quality genes out of 12 genes input (8.33%).&quot;</code></pre>
<pre class="sourceCode r" id="cb17"><code class="sourceCode r"><div class="sourceLine" id="cb17-1" data-line-number="1">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##       A  B  C  D  E
## ACTB 63 19 59 14  0
## a    36 47 28 21  1
## b     5  6  1 13  5
## c     1 11  1  9 12
## d    84  0  0  8 14
## e     8  1 20  3  0
## f     9 35 50  7  9
## g     5  0  7  1  8
## h     7  1  5  8  8
## i    17  1 17 23 47
## j     5  3  3 12 18</code></pre>
</div>
<div id="remove-low-quality-samples" class="section level3">
<h3>Remove low quality samples</h3>
<p>It is often the case that some samples from sequencing experiments are of low quality, in some cases due to issues during the sample preperation stage. Due to the fact that these samples represent a high level of technical noise, it is often desirable to remove these before downstream analysis which is facilitated by the <code>detectLowQualityCells</code> function. The function achieves this using two methods.First, the mincount argument allows detection of samples whose sum across all genes is &gt; mincount. Second, we utilize a house keeping gene and assume its log2 expression to be normally distributed when considering all samples. We then detect samples where the probability of the expression for the house keeping gene in that sample is greater than the quantile.cut argument. The function returns a logical vector indicating cells that “pass” both of the tests that can subsequently be used to subset them from the counts matrix. A message is issued indicating the number of low quality cells detected and the percentage of low quality cells in the dataset.</p>
<pre class="sourceCode r" id="cb19"><code class="sourceCode r"><div class="sourceLine" id="cb19-1" data-line-number="1">lqc &lt;-<span class="st"> </span><span class="kw">detectLowQualityCells</span>(</div>
<div class="sourceLine" id="cb19-2" data-line-number="2">  counts,</div>
<div class="sourceLine" id="cb19-3" data-line-number="3">  <span class="dt">geneName =</span> <span class="st">&quot;ACTB&quot;</span>,</div>
<div class="sourceLine" id="cb19-4" data-line-number="4">  <span class="dt">mincount =</span> <span class="dv">30</span>,</div>
<div class="sourceLine" id="cb19-5" data-line-number="5">  <span class="dt">quantileCut =</span> <span class="fl">0.01</span></div>
<div class="sourceLine" id="cb19-6" data-line-number="6">)</div></code></pre>
<pre><code>## [1] &quot;Detected 2 low quality cells out of 11 cells input (18.18%).&quot;</code></pre>
<pre class="sourceCode r" id="cb21"><code class="sourceCode r"><div class="sourceLine" id="cb21-1" data-line-number="1">counts &lt;-<span class="st"> </span>counts[, lqc]</div>
<div class="sourceLine" id="cb21-2" data-line-number="2">countsERCC &lt;-<span class="st"> </span>countsERCC[, lqc]</div>
<div class="sourceLine" id="cb21-3" data-line-number="3"></div>
<div class="sourceLine" id="cb21-4" data-line-number="4">counts[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##       A  B  C  D  F
## ACTB 63 19 59 14 75
## a    36 47 28 21  1
## b     5  6  1 13  2
## c     1 11  1  9  0
## d    84  0  0  8  0
## e     8  1 20  3  2
## f     9 35 50  7  2
## g     5  0  7  1 28
## h     7  1  5  8 14
## i    17  1 17 23  5
## j     5  3  3 12  0</code></pre>
<pre class="sourceCode r" id="cb23"><code class="sourceCode r"><div class="sourceLine" id="cb23-1" data-line-number="1">countsERCC[, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</div></code></pre>
<pre><code>##        A B C D F
## ERCC-1 1 0 0 2 2</code></pre>
</div>
<div id="coerce-to-matrix" class="section level3">
<h3>Coerce to matrix</h3>
<p>Due to the fact that, when origionally loaded, the counts data includes a character column (HGN) they are typically stored in a data.frame. Despite this, downstream processing is usually easier when the data are in a matrix data structure. The <code>convertCountsToMatrix</code> coreces the counts data to a matrix.</p>
<pre class="sourceCode r" id="cb25"><code class="sourceCode r"><div class="sourceLine" id="cb25-1" data-line-number="1">counts &lt;-<span class="st"> </span><span class="kw">convertCountsToMatrix</span>(counts)</div>
<div class="sourceLine" id="cb25-2" data-line-number="2">countsERCC &lt;-<span class="st"> </span><span class="kw">convertCountsToMatrix</span>(countsERCC)</div>
<div class="sourceLine" id="cb25-3" data-line-number="3"></div>
<div class="sourceLine" id="cb25-4" data-line-number="4"><span class="kw">class</span>(counts)</div></code></pre>
<pre><code>## [1] &quot;matrix&quot;</code></pre>
<pre class="sourceCode r" id="cb27"><code class="sourceCode r"><div class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">class</span>(countsERCC)</div></code></pre>
<pre><code>## [1] &quot;matrix&quot;</code></pre>
</div>
<div id="counts-filtering-summary" class="section level3">
<h3>Counts filtering summary</h3>
<p>In summary the whole workflow can be reduced to the pipeline shown below. Note that order of operations is important.</p>
<pre class="sourceCode r" id="cb29"><code class="sourceCode r"><div class="sourceLine" id="cb29-1" data-line-number="1">counts &lt;-<span class="st"> </span>testingCounts <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-2" data-line-number="2"><span class="st">  </span><span class="kw">moveGenesToRownames</span>() <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-3" data-line-number="3"><span class="st">  </span><span class="kw">removeHTSEQsuffix</span>()</div>
<div class="sourceLine" id="cb29-4" data-line-number="4"></div>
<div class="sourceLine" id="cb29-5" data-line-number="5">ercc &lt;-<span class="st"> </span><span class="kw">detectERCCreads</span>(counts, <span class="dt">warn =</span> <span class="ot">FALSE</span>)</div>
<div class="sourceLine" id="cb29-6" data-line-number="6"></div>
<div class="sourceLine" id="cb29-7" data-line-number="7">filtered.counts &lt;-<span class="st"> </span>counts[<span class="op">!</span>ercc, ] <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-8" data-line-number="8"><span class="st">  </span><span class="kw">subset</span>(<span class="op">!</span><span class="kw">detectNonGenes</span>(.)) <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-9" data-line-number="9"><span class="st">  </span><span class="kw">subset</span>(<span class="kw">detectLowQualityGenes</span>(., <span class="dv">18</span>)) <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-10" data-line-number="10"><span class="st">  </span><span class="kw">select</span>(<span class="kw">which</span>(</div>
<div class="sourceLine" id="cb29-11" data-line-number="11">    <span class="kw">detectLowQualityCells</span>(</div>
<div class="sourceLine" id="cb29-12" data-line-number="12">      ., <span class="dt">geneName =</span> <span class="st">&quot;ACTB&quot;</span>, <span class="dt">mincount =</span> <span class="dv">30</span>, <span class="dt">quantileCut =</span> <span class="fl">0.01</span></div>
<div class="sourceLine" id="cb29-13" data-line-number="13">    )</div>
<div class="sourceLine" id="cb29-14" data-line-number="14">  )) <span class="op">%&gt;%</span></div>
<div class="sourceLine" id="cb29-15" data-line-number="15"><span class="st">  </span><span class="kw">convertCountsToMatrix</span>(.)</div></code></pre>
<pre><code>## [1] &quot;Detected 1 low quality genes out of 12 genes input (8.33%).&quot;
## [1] &quot;Detected 2 low quality cells out of 11 cells input (18.18%).&quot;</code></pre>
<pre class="sourceCode r" id="cb31"><code class="sourceCode r"><div class="sourceLine" id="cb31-1" data-line-number="1">countsERCC &lt;-<span class="st"> </span><span class="kw">convertCountsToMatrix</span>(counts[ercc, <span class="kw">colnames</span>(filtered.counts)])</div></code></pre>
</div>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<p>In addition to the functions already mentioned, there is a subset of functions devoted to helping with metadata. Since the metadata tends to be quite specific depending on the project, it is difficult to write generalized functions to help with metadata. Despite this, some things are assumed to be shared between all projects. The functions below leverage a sample naming convention to annotate various metadata. In the future, these functions may be updated to be more useful in projects using a different sample naming convention. A example of a metadata file is included as “testingMeta” and includes sample names with the typical format. The naming convention for samples is as follows:</p>
<p>Sample names should follow: (s|m)\\.platename\\.platePosition. platePosition is in the form row and column without a space where row is a LETTER (A-H) and column is a number (1-12).</p>
<pre class="sourceCode r" id="cb32"><code class="sourceCode r"><div class="sourceLine" id="cb32-1" data-line-number="1">testingMeta</div></code></pre>
<pre><code>## # A tibble: 6 x 1
##   sample        
##   &lt;chr&gt;         
## 1 s.NJB00201.B01
## 2 s.NJB00201.A07
## 3 s.NJB00201.F03
## 4 s.NJB00201.F07
## 5 m.NJB00204.A09
## 6 m.NJB00204.B05</code></pre>
<div id="annotate-plate" class="section level3">
<h3>Annotate Plate</h3>
<p>Uses the standard sample naming nomenclature to add plate row to metadata. Returns the metadata tibble with an additional column indicating plate name.</p>
<pre class="sourceCode r" id="cb34"><code class="sourceCode r"><div class="sourceLine" id="cb34-1" data-line-number="1">meta &lt;-<span class="st"> </span><span class="kw">annotatePlate</span>(testingMeta)</div>
<div class="sourceLine" id="cb34-2" data-line-number="2">meta</div></code></pre>
<pre><code>## # A tibble: 6 x 2
##   sample         plate   
##   &lt;chr&gt;          &lt;chr&gt;   
## 1 s.NJB00201.B01 NJB00201
## 2 s.NJB00201.A07 NJB00201
## 3 s.NJB00201.F03 NJB00201
## 4 s.NJB00201.F07 NJB00201
## 5 m.NJB00204.A09 NJB00204
## 6 m.NJB00204.B05 NJB00204</code></pre>
</div>
<div id="annotate-row" class="section level3">
<h3>Annotate Row</h3>
<p>Uses the standard sample naming nomenclature to add plate row to metadata. Returns the metadata tibble with an additional column indicating row as a numeric value.</p>
<pre class="sourceCode r" id="cb36"><code class="sourceCode r"><div class="sourceLine" id="cb36-1" data-line-number="1">meta &lt;-<span class="st"> </span><span class="kw">annotateRow</span>(meta)</div>
<div class="sourceLine" id="cb36-2" data-line-number="2">meta</div></code></pre>
<pre><code>## # A tibble: 6 x 3
##   sample         plate      row
##   &lt;chr&gt;          &lt;chr&gt;    &lt;int&gt;
## 1 s.NJB00201.B01 NJB00201     2
## 2 s.NJB00201.A07 NJB00201     1
## 3 s.NJB00201.F03 NJB00201     6
## 4 s.NJB00201.F07 NJB00201     6
## 5 m.NJB00204.A09 NJB00204     1
## 6 m.NJB00204.B05 NJB00204     2</code></pre>
</div>
<div id="annotate-column" class="section level3">
<h3>Annotate column</h3>
<p>Uses the standard sample naming nomenclature to add plate column to metadata. Returns the metadata tibble with an additional column indicating column as a numeric value.</p>
<pre class="sourceCode r" id="cb38"><code class="sourceCode r"><div class="sourceLine" id="cb38-1" data-line-number="1">meta &lt;-<span class="st"> </span><span class="kw">annotateColumn</span>(meta)</div>
<div class="sourceLine" id="cb38-2" data-line-number="2">meta</div></code></pre>
<pre><code>## # A tibble: 6 x 4
##   sample         plate      row column
##   &lt;chr&gt;          &lt;chr&gt;    &lt;int&gt;  &lt;int&gt;
## 1 s.NJB00201.B01 NJB00201     2      1
## 2 s.NJB00201.A07 NJB00201     1      7
## 3 s.NJB00201.F03 NJB00201     6      3
## 4 s.NJB00201.F07 NJB00201     6      7
## 5 m.NJB00204.A09 NJB00204     1      9
## 6 m.NJB00204.B05 NJB00204     2      5</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
